File: src/p1xxx/p1717.py
from leetcode_prelude import *


# Problem 1717
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        a = "a"
        b = "b"

        # make sure x > y
        if x < y:
            x, y = y, x
            a, b = b, a

        res = 0

        # first, remove all possible ab (more points)

        stack = []
        for ch in s:
            if stack and stack[-1] == a and ch == b:
                stack.pop()
                res += x
            else:
                stack.append(ch)

        # second, remove all ba (less points)

        stack2 = []
        for ch in stack:
            if stack2 and stack2[-1] == b and ch == a:
                stack2.pop()
                res += y
            else:
                stack2.append(ch)

        return res

File: src/p1xxx/p1717.rs
// Problem 1717
impl Solution {
    pub fn maximum_gain(s: String, x: i32, y: i32) -> i32 {
        // make sure x >= y. otherwise, swap a and b so are x and y
        let (a, b, x, y) = if x >= y {
            ('a', 'b', x, y)
        } else {
            ('b', 'a', y, x)
        };

        let mut score = 0;
        let mut stack = vec![];
        for ch in s.chars() {
            if ch == b && stack.last().copied() == Some(a) {
                stack.pop();
                score += x;
            } else {
                stack.push(ch);
            }
        }

        let mut stack2 = vec![];
        for ch in stack.into_iter() {
            if ch == a && stack2.last().copied() == Some(b) {
                stack2.pop();
                score += y;
            } else {
                stack2.push(ch);
            }
        }

        score
    }
}
